Add OLMo model family

2015aroras opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

This PR adds the OLMo model family to transformers. A base `OLMoModel` and a casual LM `OLMoForCausalLM` are implemented. The models are already present in HF Hub (e.g. [allenai/OLMo-7B](https://huggingface.co/allenai/OLMo-7B)), and this implementation is compatible with the checkpoints in the Hub.

UPDATE: The current version of the PR is not compatible with the old OLMo models in HF Hub. New models have been uploaded to HF Hub (e.g. [allenai/OLMo-7B-hf](https://huggingface.co/allenai/OLMo-7B-hf)) to support this PR.

<!-- Remove if not applicable -->

Fixes #29885


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

2015aroras committed:
Add OLMo using add-new-model-like with Llama

2015aroras committed:
Fix incorrect tokenizer for OLMo

2015aroras committed:
Copy-paste relevant OLMo methods and their imports

2015aroras committed:
Add OLMo config

2015aroras committed:
Modify OLMo config to follow HF conventions

2015aroras committed:
Remove unneeded Llama code from OLMo model

2015aroras committed:
Add ability for OLMo model to output attentions

2015aroras committed:
Add OLMoPreTrainedModel and OLMoModel

2015aroras committed:
Add OLMoForCausalLM

2015aroras committed:
Minor fixes to OLMo model for style and missing functions

2015aroras committed:
Implement OLMo tokenizer

2015aroras committed:
Implement OLMo to HF conversion script

2015aroras committed:
Add tests for OLMo model

2015aroras committed:
Add tests for OLMo fast tokenizer

2015aroras committed:
Add auto-generated dummy objects

2015aroras committed:
Remove unimplemented OLMo classes from auto and init classes and re-format

2015aroras committed:
Add README and associated auto-generated files

2015aroras committed:
Use OLMo names for common properties

2015aroras committed:
Run make fixup

2015aroras committed:
Remove `|` from OLMo typing

2015aroras committed:
Remove unneeded tokenization_olmo.py

2015aroras commented:
@ArthurZucker and @younesbelkada for general guidance and/or review, since this is a PyTorch text model.

ArthurZucker reviewed (commented):
Hey! Thanks for submitting the PR!
The code seems in a very research state, let's try to get it to `transformers` philosophy. 
My main query would be: what are the main differences with `MixtralMoe` ? (Llama + MoE) ? Only include differences for trained checkpoints (I see both alibi and rotary, all of them use both? Or only one was kept?) 

The fast tokenizer also looks very similar to GptNeoX / Bloom one, any reason to have a new one?

2015aroras commented:
> Hey! Thanks for submitting the PR! The code seems in a very research state, let's try to get it to `transformers` philosophy. My main query would be: what are the main differences with `MixtralMoe` ? (Llama + MoE) ? Only include differences for trained checkpoints (I see both alibi and rotary, all of them use both? Or only one was kept?)
> 
> The fast tokenizer also looks very similar to GptNeoX / Bloom one, any reason to have a new one?

In a general sense, I tried to stay as faithful to OLMo as I could while following general `transformers` principles (all model code in 1 file, use "Copied from" with re-usable logic, etc.).

Regarding tokenization, OLMo adds an eos token to the end of inputs that don't have them. The GptNeoX and Bloom ones don't. Llama's tokenizer does, but it also has a slow tokenizer and I couldn't figure out how to convert the fast OLMo tokenizer to a slow `sentencepiece` one without reinventing the wheel. My current implementation copies a lot from the Llama fast tokenizer.

In terms of the model, OLMo doesn't have MoE whereas `MixtralMoe` does. I would have to figure out if I could use `MixtralMoe` while avoiding the MoE functionality. OLMo is more directly map-able to Llama; the model architectures mostly map 1 to 1, and the main discrepancies in output come from the 2 models using different precision for some operations.

The official OLMo models use rope and at most 1 of rope and alibi is used in a run (i.e. no alibi), but we hope to release checkpoints from our experiments at some point. Some of those would use alibi. I think it would be nice if transformers could support those checkpoints too.

@dirkgr FYI

ArthurZucker commented:
Hey! Sorry I might have misunderstood (thought I saw some MoE things here). 
Alright when I mean research I mean there are a lot of things that we never do in transformers. 
1. Tokenizer, we can update bloom / Gpneox to support this option as IMO it makes a lot of sense! No need for a slow one 😉 
2. The modeling code does not seem to include a lot of `# Copied from`, would recommend you to look at `Gemma` modeling for a good example! Basically we want to avoid all the autocasting code when it is not necessary. we do checkpointing a bit differently and we want to have the least number of code paths as possible. 
3. For the config, all of the different enums for layer norms and act fn, we already have the `from ...activations import ACT2FN
` and the `ALL_LAYER_NORMS` which should be used. 
4. `BufferCache` is not really something we use, and I am not sure we had issue with the community after the last Llama fixes!
5. `k_norm` and `q_norm` are used in `Persimmon` so a good place to copy from as well!
6. `_scaled_dot_product_attention` should follow the classing we use! So xxxxSdpaAttention
7. if `clip_qkv` is always true let's not have a if else!
8. attention mask logic should use the `_update_causal_mask` from Llama and have a similar logic
9. We don't use `reset_parameters` or `device` let's try to get closer to Llama!
10. We don't incude `get_fsdp_wrap_policy` and fsdp specifics in transformers
11. Let's add some generation tests!

2015aroras commented:
Just want to clarify my thoughts before I try making changes, so I'm expressing my understanding below. Please correct me where I am wrong.

Regarding the tokenizer, it sounds like (if possible) I should update one of the existing tokenizers and then use that tokenizer directly (similar to how `mamba` appears to use `GPTNeoXTokenizerFast`).

The comments regarding modeling code suggest that I should aim to re-use existing code using `#Copied from` and any existing functions. More generally, I should aim to be more "transformers-like" than faithful to the original code. I should try to make the code run like OLMo (produce the same output) but look like transformers.

----

Regarding copying existing code vs being faithful, what should I do when some operation (say, RoPE) exists in transformers but is performed in a different precision (e.g. float32 vs bfloat16) in OLMo compared to the transformer code? Do I use the transformers code and disregard the "small" error, or write separate code that matches the original OLMo behavior?

ArthurZucker commented:
> Assuming it were possible, the ideal solution would be for me to use some existing model and convert OLMo checkpoints to work with that model.

Yes! This way it's easier for all the community to use the model, and keeps the diff very clear and clean! 

For the tokenizer, if `GPTNeoXTokenizerFast` does not have `add_eos` or `add_bos`, we can add it but in a backward compatible way! 

> The comments regarding modeling code suggest that I should aim to re-use existing code using #Copied from and any existing functions. More generally, I should aim to be more "transformers-like" than faithful to the original code. I should try to make the code run like OLMo (produce the same output) but look like transformers.

Exactly 💯

2015aroras committed:
Revert model, config and converter to add-new-model-like Llama

2015aroras committed:
Move logic for adding bos/eos token into GPTNeoxTokenizerFast

2015aroras committed:
Change OLMoConfig defaults to match OLMo-7B

2015aroras committed:
Use GPTNeoXToknizerFast in OLMo tokenizer tests

2015aroras committed:
Modify auto-generated OLMoModelTests to work for OLMo

2015aroras committed:
Add non-parametric layer norm OLMoLayerNorm

2015aroras committed:
Update weight conversion script for OLMo

2015aroras committed:
Fix __init__ and auto structure for OLMo

2015aroras committed:
Fix errors from make fixup

2015aroras committed:
Remove OLMoTokenizerFast from documentation

2015aroras commented:
Due to the above discussion, I have redone the implementation. The key points are:

- For tokenizer, OLMo now uses `GPTNeoXTransformerFast`. I have modified the tokenizer to make adding `bos_token` or `eos_token` an option. Unfortunately, that tokenizer doesn't come with any tests...
- For model, I have reverted mostly to the code generated from [add-new-model-like with Llama](https://github.com/huggingface/transformers/pull/29890/commits/0afbd4a35925394ee4c9010e6bf985d87adac764). The main changes are changing the norm into the non-parametric version used by OLMo and removing auto-generated model classes like `OLMoForQuestionAnswering`. Should I add those classes back (I have not spent time trying to understand those classes yet)?

The existing checkpoints on HF Hub are not compatible with this implementation, so integration tests are currently broken. I can work on fixing that once we are happy with the approach/implementation. On a 1B checkpoint that I converted using the updated conversion script, the maximal logit difference is `~2.5e-5`.

ArthurZucker commented:
Thanks! Pinging @molbap for the review!

2015aroras commented:
@molbap Gentle ping

molbap reviewed (commented):
Hey, got a review out! I thnk the code is in a good state- I left some pointers around the places where it seems to break. If there is no change in the way `causal_mask` is used w.r.t. llama/gemma, I would advise to use the original `_update_causal_mask` method! 

Let me know when integration + tokenizers test are green on your end and I'll do another pass!

2015aroras committed:
Merge remote-tracking branch 'upstream/main' into add-olmo-implementation

2015aroras committed:
Add missing 'Copied from' for OLMoModel._update_causal_mask

2015aroras committed:
Run make fix-copies

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1334 @@
+    # KV cache is used. This is an issue for torch.compile which then recaptures cudagraphs at each decode steps due to the dynamic shapes.
+    # (`recording cudagraph tree for symint key 13`, etc.), which is VERY slow. A workaround is `@torch.compiler.disable`, but this prevents using
+    # `fullgraph=True`. See more context in https://github.com/huggingface/transformers/pull/29114
+    def _update_causal_mask(self, attention_mask, input_tensor):
```
https://github.com/huggingface/transformers/pull/29890/commits/98db6a929eb7170846e34ac93de183cc457218ae Sorry, I missed a 'Copied from' here. `_update_causal_mask ` this is supposed to be identical to Llama. The only part of the OLMo model code that is supposed to be different is the non-parametric layer norm.

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1334 @@
+        if position_ids is None:
+            position_ids = cache_position.unsqueeze(0)
+
+        causal_mask = self._update_causal_mask(attention_mask, inputs_embeds, cache_position)
```
Refer to https://github.com/huggingface/transformers/pull/29890#discussion_r1554288624

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1334 @@
+            elif attention_mask.dim() == 4:
+                mask_shape = attention_mask.shape
+                mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+                causal_mask[: mask_shape[0], : mask_shape[1], : mask_shape[2], : mask_shape[3]] = mask_slice
```
Refer to https://github.com/huggingface/transformers/pull/29890#discussion_r1554288624

2015aroras commented on tests/models/olmo/test_tokenization_olmo.py:
I believe that one test from `Mixin` with `['allenai/OLMo-1B']` instead of `'allenai/OLMo-1B'` was a bug with the test instead of my code. After merging an updated main, the error is gone.

I currently have `1 failed, 91 passed, 4 skipped, 7 warnings` when I run. The one failure (`OLMoTokenizationTest::test_tokenizer_integration`) is because the current OLMo-1B tokenizer config doesn't have `"add_eos_token": true`, since it is not yet updated. Are you seeing more failures?

2015aroras commented on tests/models/olmo/test_modeling_olmo.py:
```diff
@@ -0,0 +1,455 @@
+            "Simply put, the theory of relativity states that ",
+            "My favorite all time favorite condiment is ketchup.",
+        ]
+        tokenizer = AutoTokenizer.from_pretrained("allenai/OLMo-hf", pad_token="</s>", padding_side="right")
```
Yep, I haven't sorted out the integration tests yet so that is just the auto-generated `Llama-hf` -> `OLMo-hf`.

2015aroras committed:
Rearrange string replacements in OLMoForCausalLM Copied from

2015aroras committed:
Move OLMo and Llama CausalLM.forward example into global constants

2015aroras committed:
Fix OLMO_GENERATION_EXAMPLE doc string typo

2015aroras commented:
Do you have any recommendations for how to transition our checkpoints in HF Hub from the old style that relied on remote code to the new checkpoints from this PR? I can accordingly work on updating HF Hub and fixing the integration tests.

Other than integration tests that rely on a HF Hub checkpoint, everything else should be passing now.

molbap commented:
@2015aroras For checkpoint adaptation, I'd take a look at [conversion scripts for gemma ](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma/convert_gemma_weights_to_hf.py). Overall not too far from what you are doing, load from config with correctly shaped parameters. Usually how I go about it is running the two implementations side-by-side in 2 different debuggers and look where the activations start to diverge, fixing the conversion script as I go.
For the tests, with the rebase it looks better now indeed! When the integration tests run we'll be able to ask for a core maintainer final approval :)

2015aroras commented:
Sorry, I should have been more specific about my concerns regarding transitioning our checkpoints in HF hub. My problem is that if I load the new-style OLMo checkpoints (i.e. OLMo checkpoints converted using `convert_olmo_weights_to_hf.py`), then it would break HF OLMo for people without the code from this PR. OLMo is already consumable using `OLMoForCausalLM.from_pretrained("OLMo-7B", trust_remote_code=True)`, but once I upload the new-style checkpoint their code will break until they update the HF library. How can I upload the new-style OLMo checkpoints without breaking people that are using older `transformers` versions?

It looks like Llama got around this issue by having separate `meta-llama/Llama-2-7b` and `meta-llama/Llama-2-7b-hf`, where the latter is "converted for the Hugging Face Transformers format". Is this approach acceptable?

molbap commented:
Ah, I understand better now! Yes, we usually recommend to go for one repo on the hub = one model. In that case, it would be acceptable to add a -hf suffix to the new checkpoints on the hub!

2015aroras committed:
Add option for qkv clipping to OLMo

2015aroras committed:
Rearrange OLMoConfig kwargs in convert_olmo_weights_to_hf

2015aroras committed:
Add clip_qkv to OLMoConfig in convert_olmo_weights_to_hf

2015aroras committed:
Fix OLMo tokenization bug using conversion script

2015aroras committed:
Keep model in full precision after conversion

2015aroras committed:
Do not add eos token automatically

2015aroras committed:
Update references to OLMo model in HF Hub

2015aroras committed:
Do not add eos token during encoding by default

2015aroras committed:
Fix Llama generation example

2015aroras committed:
Run make fixup

2015aroras committed:
OLMo 7B integration test fix

2015aroras committed:
Merge branch 'main' into add-olmo-implementation

2015aroras committed:
Remove unneeded special case for OLMoConfig

2015aroras committed:
OLMo 7B Twin 2T integration test fix

2015aroras committed:
Fix test_model_7b_greedy_generation

2015aroras committed:
Remove test_compile_static_cache

2015aroras committed:
Fix OLMo and Llama generation example

2015aroras committed:
Run make fixup

2015aroras commented:
I have now put up "-hf" models as _private_ on the allenai HF Hub organization. Once I get the all clear internally, we will make them public. All the integration tests pass for me locally (they fail here because the repo is private).

We are hoping to release some new model checkpoints using the same architecture early next week, so it would be great if we could get this PR in quickly. Then we would be able to release models natively integrated with HF along with the rest of the release!

2015aroras commented on src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:
```diff
@@ -73,9 +73,15 @@ class GPTNeoXTokenizerFast(PreTrainedTokenizerFast):
             other word. (GPTNeoX tokenizer detect beginning of words by the preceding space).
+        add_bos_token (`bool`, *optional*, defaults to `False`):
+            Whether or not to add a `bos_token` at the start of sequences.
+        add_eos_token (`bool`, *optional*, defaults to `False`):
```
I have switched OLMo to not add the eos token by default, since that is its behavior at inference time. Consequently the changes to this tokenizer are no longer strictly needed. There is no harm to keep them though, and the largest parts of the changes are copied from Llama anyways.

molbap replied:
yeah I don't think that's an issue!

molbap commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -1110,6 +1111,26 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position, curr
         return causal_mask
 
 
+LLAMA_GENERATION_EXAMPLE = r"""
```
cc @ydshieh is this alright for the doctest?

molbap commented on src/transformers/models/olmo/configuration_olmo.py:
```diff
@@ -0,0 +1,190 @@
+        initializer_range=0.02,
+        use_cache=True,
+        pad_token_id=1,
+        bos_token_id=None,
```
is it wanted that default config do not set the `bos_token_id`?

2015aroras replied:
OLMo's config class (in the original Github) doesn't have a bos token and so no OLMo checkpoints use bos tokens to my knowledge. Thus setting no `bos_token_id` is intentional.

molbap commented:
I think the PR is in a good state. Thanks for adding this! pinging @ArthurZucker for a final review :)

2015aroras committed:
Revert "OLMo 7B integration test fix"

This reverts commit 4df56a4b150681bfa559846f40e9b7b7f97d7908.

2015aroras committed:
Revert "OLMo 7B Twin 2T integration test fix"

This reverts commit 9ff65a4a294ace89ab047b793ca55e623a9ceefc.

2015aroras committed:
Ungate 7B integration tests and fix greedy generation test

2015aroras committed:
Add retries for flaky test_eager_matches_sdpa_generate

2015aroras committed:
Fix output of doc example for OLMoForCausalLM.forward

2015aroras committed:
Downsize OLMo doc test for OLMoForCausalLM.forward to 1B model

2015aroras committed:
Try fix incorrect characters in OLMoForCausalLM.forward doct test

2015aroras committed:
Try fix incorrect characters in OLMoForCausalLM.forward doc test using end quotes

2015aroras commented:
@molbap @ArthurZucker All checks are passing! Now it's just a matter of reviews

ArthurZucker commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -174,6 +174,7 @@
         ("nllb-moe", "NllbMoeConfig"),
         ("nougat", "VisionEncoderDecoderConfig"),
         ("nystromformer", "NystromformerConfig"),
+        ("olmo", "OLMoConfig"),
```
```suggestion
        ("olmo", "OlmoConfig"),
```
we need camel casing on all classes

2015aroras replied:
[c7f4e74](https://github.com/huggingface/transformers/pull/29890/commits/c7f4e7489836d43b1d6a34cf47eb1de1120d1961)

ArthurZucker commented on src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:
LGTM

ArthurZucker commented on src/transformers/models/llama/modeling_llama.py:
let's not modify an unrelated file as important as Llama! Remove the copied from and for the forward only! / use `# Ignore copy`

2015aroras replied:
[826c91b](https://github.com/huggingface/transformers/pull/29890/commits/826c91ba7106eacfb94de8dfd1f7870559e08a95)

ArthurZucker commented on src/transformers/models/olmo/configuration_olmo.py:
```diff
@@ -0,0 +1,190 @@
+            Beginning of stream token id.
+        eos_token_id (`int`, *optional*, defaults to 50279):
+            End of stream token id.
+        pretraining_tp (`int`, *optional*, defaults to 1):
+            Experimental feature. Tensor parallelism rank used during pretraining. Please refer to [this
+            document](https://huggingface.co/docs/transformers/main/perf_train_gpu_many#tensor-parallelism) to understand more about it. This value is
+            necessary to ensure exact reproducibility of the pretraining results. Please refer to [this
+            issue](https://github.com/pytorch/pytorch/issues/76232).
```
Let's get rid of that no?

ArthurZucker commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+_CONFIG_FOR_DOC = "OLMoConfig"
+
+
+def _get_unpad_data(attention_mask):
```
this is copied from

2015aroras replied:
[3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572)

ArthurZucker commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+        return cos, sin
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
```
```suggestion
# Copied from transformers.models.llama.modeling_llama.rotate_half
def rotate_half(x):
    """Rotates half the hidden dims of the input."""
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)

```

2015aroras replied:
[3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572)

ArthurZucker reviewed (commented):
Good state!
- camel casing needs to be used on a ll classes!
- the main diff is `clamping`, it should not be an if else, because else it's a llama model 🤗

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
+
+        if self.config.clip_qkv is not None:
+            query_states.clamp_(min=-self.config.clip_qkv, max=self.config.clip_qkv)
+            key_states.clamp_(min=-self.config.clip_qkv, max=self.config.clip_qkv)
+            value_states.clamp_(min=-self.config.clip_qkv, max=self.config.clip_qkv)
```
It's not used for the old OLMo model that is currently on HF Hub, but it is always used with the new OLMo model that we are releasing soon. We're going to have both on HF Hub in the long term, so I don't see a way around this.

2015aroras replied:
Just to be clear, the other main diff from Llama is the non-parametric layer norm.

ArthurZucker replied:
Mmmm got it

2015aroras committed:
Remove pretraining_tp from OLMo config and model

2015aroras committed:
Add missing 'Copied from' instances

2015aroras committed:
Remove unneeded causal_mask from OLMoModel

2015aroras committed:
Revert Llama changes

2015aroras committed:
Ignore copy for OLMoForCausalLM.forward

2015aroras committed:
Change 'OLMo' to 'Olmo' in classes

2015aroras committed:
Move minimal OLMo tokenization tests to model tests

2015aroras committed:
Add missed 'Copied from' for repeat_kv

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
[3c60cf3](https://github.com/huggingface/transformers/pull/29890/commits/3c60cf3a872de1d3f6dd338a370cb2bcc6378572) Used Llama instead, since that is the reference model for this implementation

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+        causal_mask = torch.full(
+            (config.max_position_embeddings, config.max_position_embeddings), fill_value=True, dtype=torch.bool
+        )
+        self.register_buffer("causal_mask", torch.triu(causal_mask, diagonal=1), persistent=False)
```
[a037fca](https://github.com/huggingface/transformers/pull/29890/commits/a037fcac5e098cd22443531c618d42513b51b085)

2015aroras commented on tests/models/olmo/test_tokenization_olmo.py:
[ec344c3](https://github.com/huggingface/transformers/pull/29890/commits/ec344c3af3332b9941df6dae3025f17da373fb42)

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+            attn_output = attn_output.split(self.hidden_size // self.config.pretraining_tp, dim=2)
+            o_proj_slices = self.o_proj.weight.split(self.hidden_size // self.config.pretraining_tp, dim=1)
+            attn_output = sum([F.linear(attn_output[i], o_proj_slices[i]) for i in range(self.config.pretraining_tp)])
+        else:
```
[4f062d7](https://github.com/huggingface/transformers/pull/29890/commits/4f062d76bc145aadede0cc405a0114aafe627ad7)

2015aroras commented on src/transformers/models/olmo/modeling_olmo.py:
```diff
@@ -0,0 +1,1363 @@
+            value_states = [F.linear(hidden_states, value_slices[i]) for i in range(self.config.pretraining_tp)]
+            value_states = torch.cat(value_states, dim=-1)
+
+        else:
```
[4f062d7](https://github.com/huggingface/transformers/pull/29890/commits/4f062d76bc145aadede0cc405a0114aafe627ad7)

2015aroras commented:
Addressed all PR comments, and hopefully all checks will pass.

ArthurZucker reviewed (approved):
Thanks for your great work! 🔥
